src/wms/src/main/java/org/geoserver/wms/featureinfo/FeatureInfoStylePreprocessor.java

```diff
@@ -300,13 +298,19 @@ private Rule buildDynamicGeometryRule(
         return extra;
     }
 
-    private Class<?> getTargetGeometryType(Object descriptor) {
-        if (!(descriptor instanceof GeometryDescriptor)) {
-            // we don't know what this will be, we probably evaluated a filter function
+    private Class<?> getTargetGeometryType(Expression geom) {
+        try {
+            Object descriptor = geom.evaluate(schema);
+            if (!(descriptor instanceof GeometryDescriptor)) {
+                // we don't know what this will be, we probably evaluated a filter function
+                return Geometry.class;
+            } else {
+                // see if we are dealing with a polygon
+                return ((GeometryDescriptor) descriptor).getType().getBinding();
+            }
+        } catch (IllegalArgumentException e) {
```

Comment:
I would catch Exception, different functions might fail in different ways. Besides that, we need a test.

Reply:
Thank you @aaime.  Do you have any guidance as to which existing test I can extend, please?  I find it difficult to understand the overall structure of the unit tests.

Reply:
Most of these classes are tested via integration testing. It's however possible to build also a unit test, and for this case, probably appropriate. I've asked Github copilot to come up with a test, what it spit out is not usable as a final version, but you could use it as a starting point (I've modified the output a bit to add my own comments):

```java
// copyright header here please
package org.geoserver.wms.featureinfo;

import static org.junit.Assert.*;

import org.geotools.api.feature.type.FeatureType;
import org.geotools.api.style.Style;
import org.junit.Test;

public class FeatureInfoStylePreprocessorTest {

    @Test
    public void testGeomFromWkt() throws Exception {
        // Fix this to use concatenation of a x and y
        String sld = "<StyledLayerDescriptor xmlns=\"http://www.opengis.net/sld\">" +
            "<NamedLayer><Name>Layer</Name>" +
            "<UserStyle><FeatureTypeStyle>" +
            "<Rule><TextSymbolizer>" +
            "<Geometry><ogc:Function name=\"geomFromWkt\">" +
            "<ogc:Literal>POINT(0 0)</ogc:Literal>" +
            "</ogc:Function></Geometry>" +
            "<Label><ogc:PropertyName>name</ogc:PropertyName></Label>" +
            "</TextSymbolizer></Rule>" +
            "</FeatureTypeStyle></UserStyle>" +
            "</NamedLayer></StyledLayerDescriptor>";

        // Create a mock FeatureType
        FeatureType mockFeatureType = createMockFeatureType();

        // Create an instance of FeatureInfoStylePreprocessor
        FeatureInfoStylePreprocessor preprocessor = new FeatureInfoStylePreprocessor(mockFeatureType);

        // Apply the SLD to the preprocessor
        Style style = readSLD(sld, getClass());
        style.accept(preprocessor);

        // Check the result of preprocessing is the expected one, 
       // in addition to the obvious "do not throw an exception"
    }

    // Helper method to create a mock FeatureType
    private FeatureType createMockFeatureType() {
        // Use DataUtilities to build a FeatureType that has x and y columns
    }
}
```

Reply:
I think it might belong in org.geoserver.wms.featureinfo.RenderingBasedFeatureInfoTest, if I add the special breaking style as a new SLD in \src\wms\src\test\resources\org\geoserver\wms\featureinfo ?

Reply:
Yes, you can also write an integration test there. In addition to the style, you'll have to prepare a sample data set with a x and y column then (as a property file, see what is done with the grid/repeated_lines/giantPolygon" cases then.

Reply:
Thank you, I had already proceeded with the second option (integration test) before I saw your reply, so I trust this is acceptable.